packages/components/copybutton/src/CopyButton.tsx

```diff
@@ -44,76 +75,106 @@ export interface CopyButtonProps extends CommonProps {
-}
+} & (StringValue | PromiseValue);
 
 function _CopyButton(
-  props: ExpandProps<CopyButtonProps>,
-  ref: React.Ref<HTMLDivElement>,
-) {
-  const {
+  {
+    buttonProps,
     onCopy,
     value,
     className,
+    label,
     testId = 'cf-ui-copy-button',
     tooltipText = 'Copy to clipboard',
     tooltipCopiedText = 'Copied!',
     tooltipProps,
     isDisabled = false,
+    preload,
     size = 'medium',
     ...otherProps
-  } = props;
-  const styles = getStyles({ size });
-
+  }: ExpandProps<CopyButtonProps>,
+  ref: React.Ref<HTMLDivElement>,
+) {
+  const styles = getCopyButtonStyles({ size });
+  const [isLoading, setIsLoading] = useState(false);
   const [copied, setCopied] = useState(false);
   const button = useRef<HTMLButtonElement | null>(null);
-  const handleOnCopy = useCallback(
-    (e: string) => {
-      if (onCopy) {
-        onCopy(e);
+  const resolvedValue = useRef<string | Promise<string>>(
+    typeof value === 'function' ? value() : value,
+  );
+  const timer = useRef<number | undefined>();
+
+  useEffect(() => {
+    return () => {
+      if (timer.current) {
+        clearTimeout(timer.current);
       }
+    };
+  }, []);
 
-      setCopied(true);
+  useEffect(() => {
+    const load = async () => {
+      if (preload && isPromiseLike(resolvedValue.current)) {
+        resolvedValue.current = await resolvedValue.current;
+      }
+    };
 
-      setTimeout(() => {
-        setCopied(false);
-        if (button.current) {
-          button.current.blur();
-        }
-      }, 1000);
-    },
-    [onCopy],
-  );
+    load();
+  }, [preload, value]);
+
+  const handleClick = useCallback<
+    MouseEventHandler<HTMLButtonElement>
+  >(async () => {
+    if (isPromiseLike(resolvedValue.current)) {
+      setIsLoading(true);
+      resolvedValue.current = await resolvedValue.current;
+      setIsLoading(false);
+    }
+
+    await window.navigator.clipboard.writeText(resolvedValue.current);
+    onCopy?.(resolvedValue.current);
+    setCopied(true);
+
+    timer.current = window.setTimeout(() => {
+      setCopied(false);
+      if (button.current) {
+        button.current.blur();
+      }
+    }, 1000);
+  }, [onCopy]);
 
   return (
-    <div
-      ref={ref}
-      data-test-id={testId}
-      className={cx(styles.wrapper, className)}
-      {...otherProps}
-    >
-      <CopyToClipboard text={value} onCopy={handleOnCopy}>
-        <Tooltip
-          content={copied ? tooltipCopiedText : tooltipText}
-          {...tooltipProps}
-          isDisabled={isDisabled}
-        >
-          <button
-            type="button"
-            ref={button}
-            className={cx(styles.copyButton, {
-              [styles.copyButtonDisabled]: isDisabled,
-            })}
-            aria-label={`Copy ${value} to clipboard`}
-            disabled={isDisabled}
-            aria-live="assertive"
-          >
+    <div ref={ref} data-test-id={testId} className={className} {...otherProps}>
```

Comment:
I don't remember, but I think we initially wanted to wrap most components in a root so that the `ref` passed was consistent. I've switched it to the button.